base/compiler/typeutils.jl

```diff
@@ -139,13 +139,21 @@ valid_tparam(@nospecialize(x)) = valid_typeof_tparam(typeof(x))
 
 function compatible_vatuple(a::DataType, b::DataType)
     vaa = a.parameters[end]
-    vab = a.parameters[end]
+    vab = b.parameters[end]
     if !(isvarargtype(vaa) && isvarargtype(vab))
         return isvarargtype(vaa) == isvarargtype(vab)
     end
-    (isdefined(vaa, :N) == isdefined(vab, :N)) || return false
-    !isdefined(vaa, :N) && return true
-    return vaa.N === vab.N
+    if isdefined(vaa, :N)
```

Comment:
Not really a fan of all the nesting levels. if the comparison of the isdefineds is confusing, maybe
```
if !isdefined(vaa, :N)
    return isdefined(vab, :N)
end
isdefined(vab, :N) || return false
return vaa.N == vab.N
```
?